### PR TITLE
amqp exchange and binding dynamic configuration (fixes #51)

### LINKED ISSUE #51: RabbitMQ consumer enhancements
Currently the RabbitMQ consumer(async/amqp/amqp.go) uses hardcoded configuration for the exchange/queue/bindings .

This is problematic both when patron runs on a clean Rabbit installation(as the client should provide the desired configuration), but also when the Rabbit instance has been configured, as the client will fail if a queue or exchange with a given name already exists, but has different configuration(for example if the exchange type[direct, fanout, etc] is different)

We need to update amqp, to allow the clients provide the configuration

### PR DESCRIPTION
**Pull request by harry-pap**
## Which problem is this PR solving?

Solves #51 

## Short description of the changes

A new struct Exchange containing the name and the kind of the exchange is introduced. Also, an additional parameter for the exchange-queue binding is introduced

### CODE REVIEW

**Comment by harry-pap** [comment]
I will update the consumer to accept a slice of bindings instead of a single one

File: async/amqp/amqp.go [hunk 1]
```diff
@@ -54,17 +54,47 @@ func (m *message) Nack() error {
...
-	exchange string
+	exchange Exchange
+	bindings []string
 	oo       []OptionFunc
 }
 
 // New constructor.
-func New(url, queue, exchange string, oo ...OptionFunc) (*Factory, error) {
-
+func New(url, queue string, exchange Exchange, bindings []string, oo ...OptionFunc) (*Factory, error) {
```
Comment (mantzas) [inline async/amqp/amqp.go]: we should add bindings as a option.
if no option is set then set binding to empty string.

**Review by mantzas — CHANGES_REQUESTED** [review]

File: async/amqp/amqp.go [hunk 2]
```diff
@@ -54,17 +54,47 @@ func (m *message) Nack() error {
...
+	if ex.Kind == "" {
+		return errors.New("RabbitMQ exchange type is required")
+	}
+
+	if ex.Kind != amqp.ExchangeDirect &&
+		ex.Kind != amqp.ExchangeFanout &&
+		ex.Kind != amqp.ExchangeTopic &&
+		ex.Kind != amqp.ExchangeHeaders {
+
+		return errors.New(fmt.Sprintf("RabbitMQ exchange type is invalid, one of [%s, %s, %s, %s] is required",
```
[the remark this answers is not in the document]
Comment (cpapidas) [reply async/amqp/amqp.go]: We can use fmt.Errorf("") instend of errors.New(fmt.Sprintf(.

**Comment by cpapidas** [comment]
@chpapadop could you rebase with master branch we have conflicts, thanks

**Comment by sotirispl** [comment]
FYI you have some conflicts

File: async/amqp/amqp.go [hunk 3]
```diff
@@ -54,17 +54,47 @@ func (m *message) Nack() error {
...
 }
 
-// Factory of a AMQP consumer
+// Exchange describes the Name and Kind of a RabbitMQ exchange
+type Exchange struct {
+	Name string
+	Kind string
+}
+
+func (ex Exchange) validate() error {
```
[the remark this answers is not in the document]
Comment (harry-pap) [reply async/amqp/amqp.go]: I've complied although I prefer keeping validation out of constructors

File: async/amqp/amqp.go (hunk 1 above)
Comment (harry-pap) [reply async/amqp/amqp.go]: We already have a variadic argument so I used an empty string as binding if an empty slice is provided, like you mentioned

Comment (mantzas) [reply async/amqp/amqp.go]: if you check out the option.go in the package you see how to create a option the is the last variadic argument. this is called "functional configuration"...

Comment (harry-pap) [reply async/amqp/amqp.go]: Made the bindings an OptionFunc

**Comment by harry-pap** [comment]
@cpapidas @sotirispl @mantzas 
I have addressed the code review comments, please re-review

File: async/amqp/option.go [hunk 4]
```diff
@@ -41,3 +41,14 @@ func Requeue(requeue bool) OptionFunc {
 		return nil
 	}
 }
+
+func Bindings(bindings ...string) OptionFunc {
+	return func(c *consumer) error {
+		if len(bindings) == 0 {
```
Comment (mantzas) [inline async/amqp/option.go:48]: the empty case should not be checked, if someone used the option he wants to add bindings so if none is provided an error should be return.

**Review by mantzas — CHANGES_REQUESTED** [review]

File: async/amqp/option.go (hunk 4 above)
Comment (harry-pap) [reply async/amqp/option.go:48]: Updated the function to return an error if the bindings are empty

**Review by mantzas — DISMISSED** [review]

**Review by cpapidas — DISMISSED** [review]

**Review by sotirispl — DISMISSED** [review]

**Review by sotirispl — DISMISSED** [review]

**Review by mantzas — DISMISSED** [review]

**Review by mantzas — APPROVED** [review]

**Review by mantzas — APPROVED** [review]

**Review by cpapidas — APPROVED** [review]

**Pull request merged**